src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs

```diff
@@ -20,20 +28,117 @@ impl fmt::Debug for Address {
     }
 }
 
-#[derive(Debug, Clone)]
+#[derive(Clone, Debug, Default)]
 pub struct NetworkConfig {
+    /// for detach, now it's default value 0.
+    pub index: u64,
+
     /// Host level path for the guest network interface.
     pub host_dev_name: String,
 
+    /// Guest iface name for the guest network interface.
+    pub virt_iface_name: String,
+
     /// Guest MAC address.
     pub guest_mac: Option<Address>,
 }
 
-#[derive(Debug, Clone)]
+#[derive(Clone, Debug, Default)]
 pub struct NetworkDevice {
     /// Unique identifier of the device
-    pub id: String,
+    pub device_id: String,
+
+    /// Attach count as the network device refcount
+    pub attach_count: u64,
 
     /// Network Device config info
     pub config: NetworkConfig,
 }
+
+impl NetworkDevice {
+    // new creates a NetworkDevice
+    pub fn new(device_id: String, config: &NetworkConfig) -> Self {
+        Self {
+            device_id,
+            attach_count: 0,
+            config: config.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl Device for NetworkDevice {
+    async fn attach(&mut self, h: &dyn hypervisor) -> Result<()> {
+        if self
+            .increase_attach_count()
```

Comment:
The logic here is a little bit messy. The attach could be called **MULTIPLE** times, but the logic here will return **ERROR** if we are increaing attach count that is more than zero.

I would suggest a logic like this, cc @Tim-Zhang 

```Rust
  async fn attach(&mut self, h: &dyn hypervisor) -> Result<()> {
    self.increase_attach_count().await.context(".....")?;
    if self.attach_count == 1 {
        if let Err(e) = h.add_device(DeviceType::Network(self.clone())).await {
            self.decrease_attach_count().await?;
            Err(e)
        }
    }

     Ok(());
  }
```

And the increase attach count logic should look like
```Rust
async fn increase_attach_count(&mut self) -> Result<()> {
       if let self.attach_count == MAX {
          return Err(anyhow!("device was attached too many times"));
      }
     
    self.attach_count += 1;
    Ok(())
}

Reply:
Hi @Ji-Xinyou, sorry for the change, my mistake for it.

But I don't think it's a good idea to rewrite the increase_attach_count as you said. 
And it's designed to do refcount and return true or false, then tell attach() whether do attach or do skip.
There's no need to refer`self.attach_count ==1`.